README.md
# margodux

[![Go Report Card](https://goreportcard.com/badge/github.com/cesumilo/margodux)](https://goreportcard.com/report/github.com/cesumilo/margodux)
[![codecov](https://codecov.io/gh/cesumilo/margodux/branch/master/graph/badge.svg)](https://codecov.io/gh/cesumilo/margodux)
[![GoDoc](https://godoc.org/github.com/cesumilo/margodux?status.svg)](https://godoc.org/github.com/cesumilo/margodux)

**Version:** v1.0.0

## About

Redux-like module for go.

## Motivation

I started this project because I wanted to make a video game using [pixel](https://github.com/faiface/pixel). However, most operating systems require all graphics and windowing code to be executed from the main thread of the program.

The main purpose of margodux is to provide an effective way of updating the game state without being limited by the frame rate. To do so, I needed a medium to execute code outside of the main thread but still being able to update the game state and the graphical interface.

That's how the idea came to me to build a redux-like module called `margodux`. I found some open-source implementation really close to what I was expecting, but I didn't found the perfect fit. That's why I decided to implement my own version.

## Table of contents

1. [Getting started](#getting-started)
2. [Test](#test)
3. [Changelog](#changelog)
4. [License](#license)

## Getting started

**Creating a new store**

```go
store := margodux.New()
```

**Creating a new reducer**

```go
initialState := margodux.State{"foo": "bar"}
store.Register("test", initialState, func(state State, action Action) State {
  switch action.ID {
  case "test":
    return State{"foo": "oof"}
  default:
    return initialState
  }
})
```

**Getting current state**

```go
currentState := store.GetState()
fmt.Println(currentState["test"]["foo"]) // bar
```

**Dispatching action**

```go
store.Dispatch(Action{ID: "test", Payload: nil, Err: false})

currentState := store.GetState()

fmt.Println(currentState["test"]["foo"]) // oof
```

**Dispatching async action**

```go
type TestAsyncAction struct{}

func (t *TestAsyncAction) Run(s *Store) {
  s.Dispatch(Action{ID: "test", Payload: nil, Err: false})
}

store.Dispatch(&TestAsyncAction{})

time.Sleep(1 * time.Second)

currentState := store.GetState()

fmt.Println(currentState["test"]["foo"]) // oof
```

## Test

**Run test**

```bash
go test
```

**Run test with coverage**

```bash
go test -cover
```

**Run test with coverage report and print**

```bash
go test -coverprofile=coverage.out
go tool cover -func=coverage.out
```

## Changelog

SEE CHANGELOG IN [CHANGELOG.md](CHANGELOG.md).

## License

SEE LICENSE IN [LICENSE](LICENSE).


CHANGELOG.md
# Changelog

**v1.0.0**

- Basic state machine system
- Supports async action
- Supports sync action
- Supports multi-reducers


store_test.go
/**
 * @author Guillaume Robin <guillaume@inarix.com>
 * @file Implements Store tests suite.
 * @desc Created on 2020-06-15 10:09:17 pm
 * @copyright Inarix
 */
package margodux

import (
	"testing"
	"time"
)

func TestNew(t *testing.T) {
	store := New()
	if store == nil {
		t.Error("returned nil")
	}
	if store.reducers == nil {
		t.Error("store.reducers is nil")
	}
	if store.state == nil {
		t.Error("store.state is nil")
	}
}

func TestRegister(t *testing.T) {
	store := New()
	initialState := State{"foo": "bar"}
	store.Register("test", initialState, func(state State, action Action) State {
		switch action.ID {
		case "test":
			return State{"foo": "oof"}
		default:
			return initialState
		}
	})
	if store.state["test"]["foo"] != "bar" {
		t.Error("store.Register didn't initialize state")
	}
}

func TestGetState(t *testing.T) {
	store := New()
	initialState := State{"foo": "bar"}
	store.Register("test", initialState, func(state State, action Action) State {
		switch action.ID {
		case "test":
			return State{"foo": "oof"}
		default:
			return initialState
		}
	})
	currentState := store.GetState()
	if currentState["test"]["foo"] != "bar" {
		t.Error("store.GetState didn't return current state")
	}
	if &store.state == &currentState {
		t.Error("store.GetState didn't return a copy of store's state")
	}
}

func TestDispatch(t *testing.T) {
	store := New()
	initialState := State{"foo": "bar"}
	store.Register("test", initialState, func(state State, action Action) State {
		switch action.ID {
		case "test":
			return State{"foo": "oof"}
		default:
			return initialState
		}
	})
	store.Dispatch(Action{ID: "test", Payload: nil, Err: false})
	if store.GetState()["test"]["foo"] != "oof" {
		t.Error("store.Dispatch didn't execute action")
	}
}

type TestAsyncAction struct{}

func (t *TestAsyncAction) Run(s *Store) {
	s.Dispatch(Action{ID: "test", Payload: nil, Err: false})
}

func TestDispatchAsyncAction(t *testing.T) {
	store := New()
	initialState := State{"foo": "bar"}
	store.Register("test", initialState, func(state State, action Action) State {
		switch action.ID {
		case "test":
			return State{"foo": "oof"}
		default:
			return initialState
		}
	})
	store.Dispatch(&TestAsyncAction{})
	time.Sleep(1 * time.Second)
	if store.GetState()["test"]["foo"] != "oof" {
		t.Error("store.Dispatch didn't execute action")
	}
}

func TestDispatchPanic(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Errorf("The code did not panic")
		}
	}()
	store := New()
	store.Dispatch("test")
}


engine_test.go
/**
 * @author Guillaume Robin <guillaume@inarix.com>
 * @file Implements Engine tests suite.
 * @desc Created on 2020-06-15 10:09:17 pm
 * @copyright Inarix
 */
package margodux

import (
	"testing"
	"time"
)

func TestInit(t *testing.T) {
	engine := Engine{}
	engine.Init()

	go (func() {
		for action := range engine.actions {
			switch action.id {
			case GoduxEngineQuit:
				return
			default:
				t.Error("Not sending quit signal")
				return
			}
		}
	})()

	engine.Quit()
}

func TestRun(t *testing.T) {
	engine := Engine{}
	engine.Init()

	engine.Register("app", State{"foo": "bar"}, func(in chan *Action, out chan *GlobalState) {
		for action := range in {
			switch action.id {
			case GoduxEngineQuit:
				return
			default:
				t.Error("Not sending quit signal")
				return
			}
		}
	})

	go engine.Run()

	engine.Quit()
}

type AsyncTest struct {
}

func (x AsyncTest) Run(out chan *Action, state *GlobalState) {
	out <- &Action{id: "test", err: false, payload: nil}
}

func TestDispatchAsyncAction(t *testing.T) {
	engine := Engine{}
	engine.Init()

	engine.Register("app", State{"foo": "bar"}, func(in chan *Action, out chan *GlobalState) {
		for action := range in {
			switch action.id {
			case GoduxEngineQuit:
				return
			case "test":
				out <- &GlobalState{"app": State{"foo": "oof"}}
			default:
				t.Error("Not sending quit signal")
				return
			}
		}
	})

	go engine.Run()

	test := AsyncTest{}

	engine.Dispatch(test)

	time.Sleep(1 * time.Second)

	if engine.state["app"]["foo"] != "oof" {
		t.Error("Didn't update state")
	}

	engine.Quit()
}

func TestDispatchPanic(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Errorf("The code did not panic")
		}
	}()

	engine := Engine{}
	engine.Init()

	engine.Register("app", State{"foo": "bar"}, func(in chan *Action, out chan *GlobalState) {
		for action := range in {
			switch action.id {
			case GoduxEngineQuit:
				return
			default:
				t.Error("Not sending quit signal")
				return
			}
		}
	})

	go engine.Run()

	engine.Dispatch("test")

	engine.Quit()
}


actions_test.go
/**
 * @author Guillaume Robin <robinguillaume.pro@gmail.com>
 * @file Implements margodux.actions tests suite.
 * @desc Created on 2020-06-15 9:45:12 pm
 * @copyright GNU General Public License v3.0
 */
package margodux

import "testing"

func TestQuit(t *testing.T) {
	res := Quit()
	if res.err {
		t.Error("res.err is true")
	}
	if res.id != GoduxEngineQuit {
		t.Errorf("res.id is different than: %s", GoduxEngineQuit)
	}
}


engine.go
/**
 * @author Guillaume Robin <robinguillaume.pro@gmail.com>
 * @file Engine defines a state machine.
 * @desc Created on 2020-06-15 9:30:49 pm
 * @copyright GNU General Public License v3.0
 */
package margodux

import "fmt"

// State defines a current state of a sub part of an app.
type State map[string]interface{}

// GlobalState defines a current state of an app.
type GlobalState map[string]State

// Reducer maps actions to states.
type Reducer func(in chan *Action, out chan *GlobalState)

// Engine manages an app state.
type Engine struct {
	state   GlobalState
	actions chan *Action
	updates chan *GlobalState
}

// Init initializes Engine struct.
func (e *Engine) Init() {
	e.state = make(GlobalState)
	e.actions = make(chan *Action)
	e.updates = make(chan *GlobalState)
}

// Register run a new reducer and pass it input and output channel.
func (e *Engine) Register(key string, initialState State, reducer Reducer) {
	e.state[key] = initialState
	go reducer(e.actions, e.updates)
}

// Quit sends a signal to reducers to stop.
func (e *Engine) Quit() {
	e.Dispatch(Quit())
}

// Run start state machine engine.
func (e *Engine) Run() {
	for {
		select {
		case a := <-e.actions:
			if a.id == GoduxEngineQuit {
				return
			}
		case state := <-e.updates:
			for k, v := range *state {
				e.state[k] = v
			}
		}
	}
}

// Dispatch distributes an action to reducers.
func (e *Engine) Dispatch(action interface{}) {
	switch v := action.(type) {
	case *Action:
		e.actions <- action.(*Action)
	case AsyncAction:
		go action.(AsyncAction).Run(e.actions, &e.state)
	default:
		panic(fmt.Sprintln("unknown: ", v))
	}
}


store.go
/**
 * @author Guillaume Robin <robinguillaume.pro@gmail.com>
 * @file Store defines a state machine.
 * @desc Created on 2020-06-15 9:30:49 pm
 * @copyright GNU General Public License v3.0
 */
package margodux

import (
	"fmt"
	"sync"
)

// State defines a current state of a sub part of an app.
type State map[string]interface{}

// GlobalState defines a current state of an app.
type GlobalState map[string]State

// Reducer maps actions to states.
type Reducer func(State, Action) State

// Store manages an app state.
type Store struct {
	mu       sync.RWMutex
	state    GlobalState
	reducers map[string]Reducer
}

// New create a new store.
func New() *Store {
	store := Store{
		state:    make(GlobalState),
		reducers: make(map[string]Reducer),
	}
	return &store
}

// Register run a new reducer and pass it input and output channel.
func (s *Store) Register(key string, initialState State, reducer Reducer) {
	defer s.mu.Unlock()
	s.mu.Lock()
	s.state[key] = initialState
	s.reducers[key] = reducer
}

// GetState returns a copy of the current state.
func (s *Store) GetState() GlobalState {
	defer s.mu.RUnlock()
	s.mu.RLock()
	return s.state
}

// Dispatch distributes an action to reducers.
func (s *Store) Dispatch(action interface{}) {
	switch v := action.(type) {
	case Action:
		defer s.mu.Unlock()
		s.mu.Lock()
		for key := range s.reducers {
			s.state[key] = s.reducers[key](s.state[key], action.(Action))
		}
	case AsyncAction:
		go action.(AsyncAction).Run(s)
	default:
		panic(fmt.Sprintln("unknown: ", v))
	}
}
